iree/compiler/Dialect/Flow/Transforms/DeduplicateExecutables.cpp

```diff
@@ -0,0 +1,208 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+namespace {
+
+// Replaces each usage of an entry point with its original symbol name with a
+// new symbol name.
+void replaceEntryPointUses(mlir::ModuleOp moduleOp,
+                           const DenseMap<Attribute, Attribute> &replacements) {
+  for (auto funcOp : moduleOp.getOps<mlir::FuncOp>()) {
+    funcOp.walk([&](DispatchOp dispatchOp) {
+      auto it = replacements.find(dispatchOp.entry_point());
+      if (it != replacements.end()) {
+        dispatchOp.entry_pointAttr(it->second.cast<SymbolRefAttr>());
+      }
+    });
+  }
+}
+
+bool areRegionsEquivalent(Region *lhs, Region *rhs) {
+  if (lhs->getBlocks().size() != rhs->getBlocks().size()) {
+    return false;
+  }
+
+  auto blockPairs = llvm::zip(lhs->getBlocks(), rhs->getBlocks());
+  for (auto blockPair : blockPairs) {
+    auto &lhsBlock = std::get<0>(blockPair);
+    auto &rhsBlock = std::get<1>(blockPair);
+    if (lhsBlock.getOperations().size() != rhsBlock.getOperations().size()) {
+      return false;
+    }
+
+    auto opPairs =
+        llvm::zip(lhsBlock.getOperations(), rhsBlock.getOperations());
+    for (auto opPair : opPairs) {
+      auto &lhsOp = std::get<0>(opPair);
+      auto &rhsOp = std::get<1>(opPair);
+      if (!OperationEquivalence::isEquivalentTo(
+              &lhsOp, &rhsOp, OperationEquivalence::IgnoreOperands)) {
+        return false;
+      }
+
+      // We want to check the operand _types_, but don't care if the actual
+      // operand references differ (as they live in separate modules anyway).
+      if (!std::equal(lhsOp.operand_type_begin(), lhsOp.operand_type_end(),
+                      rhsOp.operand_type_begin())) {
+        return false;
+      }
+
+      // If the operations have regions, recurse into them (depth-first).
+      if (lhsOp.getNumRegions() != rhsOp.getNumRegions()) {
+        return false;
+      }
+      auto lhsRegions = lhsOp.getRegions();
+      auto rhsRegions = rhsOp.getRegions();
+      for (int i = 0; i < lhsRegions.size(); ++i) {
+        if (!areRegionsEquivalent(&lhsRegions[i], &rhsRegions[i])) {
+          return false;
+        }
+      }
+    }
+  }
+
+  return true;
+}
+
+bool areExecutablesEquivalent(ExecutableOp lhs, ExecutableOp rhs) {
```

Comment:
try just using areRegionsEquivalent on lhsModule/rhsModule - modules have regions, and the ops (func op, dispatch entry op, etc) are just ops within that region

that'll help this generalize to other deduplication in the future as well as reduce brittleness here (if a new top level op is added to executables we don't want it to be ignored by this)

Reply:
I'd like to, but the special handling is currently needed to ignore operations having different names.

`OperationEquivalence::isEquivalentTo` checks names, then attributes:
https://github.com/llvm/llvm-project/blob/a585228027a0f025011d77916234cfba12e44309/mlir/lib/IR/OperationSupport.cpp#L598-L600
this includes the `sym_name` attribute used on our `funcOp`s.

We could change that upstream by adding a new flag to https://github.com/llvm/llvm-project/blob/a585228027a0f025011d77916234cfba12e44309/mlir/include/mlir/IR/OperationSupport.h#L863-L884, or fork to our own `isEquivalentTo` function. The `lhs->getName() != rhs->getName()` check is easy to skip, but I'm not sure how to cleanly exclude `sym_name` from `lhs->getMutableAttrDict() != rhs->getMutableAttrDict()`.

Reply:
At the point that we are setting several non-default flags for something I think it's best just to fork it -- write code that's easy to delete vs. making it more complex. You could land this now and do that as part of a follow up for when deduplication happens at other levels (HAL/VM/etc) - we don't want to copy *this* to do that, it'd be better to generalize it first.

Reply:
Writing a custom version that does what we want and using it in several places SGTM.